File: ring-android/app/src/main/java/cx/ring/tv/camera/CameraPreview.java
/*
 * Copyright (C) 2004-2021 Savoir-faire Linux Inc.
 *
 *  Author: Loïc Siret <loic.siret@savoirfairelinux.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
package cx.ring.tv.camera;

import android.content.Context;
import android.hardware.Camera;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import androidx.annotation.NonNull;

import java.io.IOException;

public class CameraPreview extends SurfaceView implements SurfaceHolder.Callback {
    private Camera mCamera;

    // Constructor that obtains context and camera
    @SuppressWarnings("deprecation")
    public CameraPreview(Context context, @NonNull Camera camera) {
        super(context);
        mCamera = camera;
        getHolder().addCallback(this);
    }

    public void stop()  {
        if (mCamera == null)
            return;
        try {
            mCamera.stopPreview();
        }  catch (Exception e) {
            // intentionally left blank
        }
        mCamera.release();
        mCamera = null;
    }

    @Override
    public void surfaceCreated(SurfaceHolder surfaceHolder) {
        if (mCamera == null)
            return;
        try {
            mCamera.setPreviewDisplay(surfaceHolder);
            mCamera.startPreview();
        } catch (Exception e) {
            // left blank for now
        }
    }

    @Override
    public void surfaceDestroyed(SurfaceHolder surfaceHolder) {
        stop();
    }

    @Override
    public void surfaceChanged(SurfaceHolder surfaceHolder, int format, int width, int height) {
        if (mCamera == null)
            return;
        try {
            mCamera.setPreviewDisplay(surfaceHolder);
            mCamera.startPreview();
        } catch (Exception e) {
            // intentionally left blank for a test
        }
    }
}

File: jami-android/app/src/main/java/cx/ring/tv/camera/CameraPreview.kt
/*
 *  Copyright (C) 2004-2025 Savoir-faire Linux Inc.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package cx.ring.tv.camera

import android.content.Context
import android.graphics.SurfaceTexture
import android.hardware.Camera
import android.util.Log
import android.view.TextureView
import java.lang.Exception

class CameraPreview(context: Context, private var mCamera: Camera?) : TextureView(context), TextureView.SurfaceTextureListener {
    private var mSurfaceTexture: SurfaceTexture? = null

    init {
        surfaceTextureListener = this
    }

    fun stop() {
        mCamera?.let { camera ->
            try {
                camera.stopPreview()
            } catch (e: Exception) {
                // intentionally left blank
            }
            camera.release()
            mCamera = null
            mSurfaceTexture?.release()
        }
        surfaceTextureListener = null
    }

    override fun onSurfaceTextureAvailable(surface: SurfaceTexture, width: Int, height: Int) {
        Log.w(TAG, "Surface texture available: $width x $height")
        startPreview(surface)
    }

    override fun onSurfaceTextureSizeChanged(surface: SurfaceTexture, width: Int, height: Int) {
        Log.w(TAG, "Surface texture changed: $width x $height")
        startPreview(surface)
    }

    override fun onSurfaceTextureDestroyed(surface: SurfaceTexture): Boolean {
        Log.w(TAG, "Surface texture destroyed")
        stop()
        mSurfaceTexture = null
        return true
    }

    override fun onSurfaceTextureUpdated(surface: SurfaceTexture) {
        // intentionally left blank
    }

    private fun startPreview(surfaceTexture: SurfaceTexture) {
        mCamera?.let { camera ->
            try {
                if (mSurfaceTexture != surfaceTexture) {
                    camera.setPreviewTexture(surfaceTexture)
                    mSurfaceTexture = surfaceTexture
                }
                camera.startPreview()
            } catch (e: Exception) {
                Log.w(TAG, "Error starting camera preview: ${e.message}")
            }
        }
    }

    companion object {
        private const val TAG = "CameraPreview"
    }
}